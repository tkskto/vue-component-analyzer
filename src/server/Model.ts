import {FORMAT} from './Constant';

class Model {
  private _resourceRoot = '.';

  private _format = FORMAT.BROWSER;

  private _outDirectory = 'out';

  private _port = '8888';

  private _isSilentMode = false;

  private _tsconfigPathMapping = new Map<string, string>();

  get resourceRoot(): string {
    return this._resourceRoot;
  }

  set resourceRoot(value: string) {
    this._resourceRoot = value;
  }

  get format(): string {
    return this._format;
  }

  set format(value: string) {
    this._format = value;
  }

  get outDirectory(): string {
    return this._outDirectory;
  }

  set outDirectory(value: string) {
    this._outDirectory = value;
  }

  get port(): string {
    return this._port;
  }

  set port(value: string) {
    this._port = value;
  }

  get isSilentMode(): boolean {
    return this._isSilentMode;
  }

  set isSilentMode(value: boolean) {
    this._isSilentMode = value;
  }

  get tsconfigPathMapping(): Map<string, string> {
    return this._tsconfigPathMapping;
  }

  set tsconfigPathMapping(paths: Map<string, string>) {
    this._tsconfigPathMapping = paths;
  }
}

export const model = new Model();
