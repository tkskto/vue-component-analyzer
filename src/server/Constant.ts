export const FORMAT = {
  BROWSER: 'browser',
  JSON: 'json',
  BOTH: 'both',
  HTML: 'html',
};
