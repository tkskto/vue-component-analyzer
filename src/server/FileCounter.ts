const cwd = process.cwd();

export class FileCounter {
  private _count: {[key: string]: number} = {};

  public add(_filename: string): void {
    const shortFilename = _filename.replace(cwd, '');

    if (Object.prototype.hasOwnProperty.call(this._count, shortFilename)) {
      this._count[shortFilename]++;
    } else {
      this._count[shortFilename] = 1;
    }
  }

  public clear(): void {
    this._count = {};
  }

  public get result(): {[key: string]: number} {
    return this._count;
  }
}

export const fileCounter = new FileCounter();
