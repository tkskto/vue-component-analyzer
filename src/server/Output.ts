import {AnalyzeReport} from '../../types';
import {Renderer} from '../client/renderer';
import {Model} from '../client/model';
import {mkdirp} from 'mkdirp';
import path from 'path';
import {writeFileSync} from 'fs';
import {render} from 'ejs';
import {fileURLToPath} from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const projectRoot = path.resolve(__dirname, '..');

export async function writeFileExtra(_filename: string, data: string) {
  const filename = _filename.endsWith('.json') ? _filename : `${_filename}/result.json`;

  try {
    await mkdirp(path.dirname(filename));

    writeFileSync(filename, data);
    console.log(`output completed ${filename}`);
  } catch (err: unknown) {
    if (err instanceof Error) {
      throw new Error(err.message);
    }
  }
}

export async function writeHTML(_filename: string, result: AnalyzeReport) {
  const filename = _filename.endsWith('.html') ? _filename : `${_filename}/result.html`;

  try {
    await mkdirp(path.dirname(filename));

    const clientModel = new Model();

    clientModel.data = result;
    clientModel.env = 'node';

    const renderer = new Renderer(clientModel);
    const htmlString = renderer.render();
    const html = render('<%- include("/views/static.ejs", {html: html}); %>', {html: htmlString}, {root: `${projectRoot}`});

    writeFileSync(filename, html);
    console.log(`output completed ${filename}`);
  } catch (err: unknown) {
    if (err instanceof Error) {
      throw new Error(err.message);
    }
  }
}
