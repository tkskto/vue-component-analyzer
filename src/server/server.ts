// for execute from npm scripts or commandline.

import AnalyzeReport = vueComponentAnalyzer.AnalyzeReport;
import express from 'express';
import {WebSocketServer} from 'ws';
import path from 'path';
import http from 'http';
import {renderFile} from 'ejs';
import opener from 'opener';
import {fileURLToPath} from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const projectRoot = path.resolve(__dirname, '..');

/**
 * start web socket server and connect when finished generate component tree.
 */
export const startServer = (port: string, json: AnalyzeReport): void => {
  const HOST = '127.0.0.1';

  // establish local web server.
  const app = express();

  // use EJS to pass variables.
  app.engine('ejs', renderFile);
  app.set('view engine', 'ejs');
  app.set('views', `${projectRoot}/views`);
  app.use(express.static(`${projectRoot}/`));

  // use same server for WebSocket Server.
  const server = http.createServer(app);
  const wss = new WebSocketServer({
    server,
  });

  app.use('/', (req: express.Request, res: express.Response) => {
    res.render('viewer', {
      mode: 'server',
      title: 'analyze report',
      enableWebSocket: true,
    });
  });

  server.on('error', (err: Error) => {
    console.log(err);

    if (server.listening) {
      server.close((serverErr: Error | undefined) => {
        console.log(serverErr);
      });
    }
  });

  server.listen(Number(port), HOST, () => {
    const address = server?.address();
    let addressPort = port;

    if (typeof address === 'string') {
      addressPort = address;
    } else if (typeof address === 'object') {
      addressPort = String(address?.port) || port;
    }

    const url = `http://${HOST}:${addressPort}/`;

    console.log(`Vue Component Analyzer is started at ${url}`);
    console.log('Use \'Ctrl+C\' to close it');

    opener(url);

    wss.on('connection', (ws: WebSocket) => {
      // pass tree data to browser.
      ws.send(JSON.stringify(json));

      ws.addEventListener('error', () => {
        console.error('Something went to wrong on web socket.');
      });
    });
  });
};
