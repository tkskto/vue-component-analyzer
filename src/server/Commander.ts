import {FORMAT} from './Constant';
import {program} from 'commander';

export const getOptions = (_argv: string[]): {
  dir: string,
  format: string,
  out: string,
  port: string,
  silent: boolean,
} => {
  program.option('--dir [dir]', 'root directory of src.', 'src');
  program.option('-f, --format [type]', 'Add the specified type of report [browser, json or both]', FORMAT.BROWSER);
  program.option('-o, --out [dir]', 'output directory (enable with setting --format option to "json" or "both")', 'out');
  program.option('-p, --port [number]', 'select a port number for the local server', '8888');
  program.option('--silent', 'do not show logs with silent flag');
  program.parse(_argv);

  return program.opts();
};
