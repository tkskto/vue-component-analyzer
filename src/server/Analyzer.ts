import FileReport = vueComponentAnalyzer.FileReport;
import {resolveFile} from './utils';
import {readFileSync, statSync} from 'fs';
import {resolve, extname} from 'path';
import {FileCounter} from './FileCounter';
import {VueComponent} from './VueComponent';
import {model} from './Model';
const cwd = process.cwd();

class Analyzer {
  private readonly _counter: FileCounter;

  constructor() {
    this._counter = new FileCounter();
  }

  /**
   * get import tree.
   * @param {string} fileName entry file name
   * @param {string[]} parents entries of file name
   * @param {boolean} isTest whether it through the test. if this doesn't exist lastModifiedTime will always fail on the GitHub Actions.
   */
  public getImportDeclarationTree = (fileName: string, parents: string[], isTest = false): FileReport => {
    const filename = resolve(cwd, fileName);
    // get filename without current working directory.
    const shortFilename = filename.replace(cwd, '');
    // get file statistic
    const stat = statSync(filename);
    // make a filename list for detecting circular dependency.
    const ancestorList: string[] = parents.concat();

    ancestorList.push(fileName);

    if (!model.isSilentMode) {
      console.log(`read ${filename}`);
    }

    // increment count of this file.
    this._counter.add(shortFilename);

    // if this file is not Vue Component file, return only filename and stat.
    if (extname(filename) === '' || extname(filename) !== '.vue') {
      return {
        name: shortFilename,
        props: '',
        size: stat.size,
        lastModifiedTime: isTest ? 0 : Number(stat.mtimeMs.toFixed(0)),
        children: [],
      };
    }

    const contents = readFileSync(filename, 'utf-8');
    const component = new VueComponent(shortFilename, contents, stat);

    try {
      // if we get, read imported file recursive.
      for (let i = 0, len = component.importDeclaration.length; i < len; i++) {
        // TODO: support other types? (value might be RegExp, or number, boolean.)
        const source = String(component.importDeclaration[i].source.value);

        if (source) {
          const nextFilename = resolveFile(source, filename);

          if (nextFilename) {
            if (parents.includes(nextFilename)) {
              console.warn(`Circular dependency detected between ${nextFilename} and ${filename}`);
            } else {
              component.addChildReport(this.getImportDeclarationTree(nextFilename, ancestorList, isTest));
            }
          }
        }
      }
    } catch (err) {
      console.error(`Something went wrong with reading ${filename}`);
      console.error(err.message);
    }

    return component.getFileReport(isTest);
  };

  get counter(): FileCounter {
    return this._counter;
  }
}

export const analyzer = new Analyzer();
