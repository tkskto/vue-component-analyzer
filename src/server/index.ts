// for execute from npm scripts or commandline.
import {AnalyzeReport} from '../../types';
import {FORMAT} from './Constant';
import {model} from './Model';
import {getOptions} from './Commander';
import {getImportDeclarationTree} from './Analyzer';
import {startServer} from './server';
import {fileCounter} from './FileCounter';
import {writeFileExtra, writeHTML} from './Output';
import {globby} from 'globby';
import path from 'path';
import {getTsconfig} from 'get-tsconfig';
import {getTsConfigPathMapping} from './utils';

// TODO: add log system
// TODO: add error system


const analyze = (entries: string[]): Promise<AnalyzeReport> => {
  return new Promise((resolve, reject) => {
    try {
      if (entries.length === 0) {
        reject(new Error('There is no entry file.'));

        return;
      }

      const entriesData = [];

      for (let i = 0, len = entries.length; i < len; i++) {
        const entryFile = entries[i];
        const children = getImportDeclarationTree(entryFile, []);

        entriesData.push(children);
      }

      resolve({
        entries: entriesData,
        count: fileCounter.result,
      });
    } catch (err) {
      console.error('something went to wrong at analyze.');
      reject(err);
    }
  });
};

export const analyzeFromCLI = async (): Promise<void> => {
  // get commander options.
  const argv = getOptions(process.argv);

  model.resourceRoot = argv.dir;
  model.format = argv.format;
  model.outDirectory = argv.out;
  model.port = argv.port;
  model.isSilentMode = argv.silent || false;

  const tsconfig = getTsconfig(argv.dir);

  if (tsconfig !== null && tsconfig.config.compilerOptions) {
    const pathMaps = getTsConfigPathMapping(tsconfig.config.compilerOptions);

    model.tsconfigPathMapping = pathMaps;
  }

  if (argv.format !== FORMAT.BROWSER && argv.format !== FORMAT.JSON && argv.format !== FORMAT.BOTH) {
    console.error(`not support ${argv.format} format.`);
  }

  console.log('start analyzing.');

  // get vue entries from dir
  const entries = await globby([argv.dir, '!**/node_modules/**'], {
    expandDirectories: {
      extensions: ['vue'],
    },
  });

  const result = await analyze(entries);

  if (argv.format === FORMAT.BOTH) {
    startServer(argv.port, result);
    await writeFileExtra(path.resolve(process.cwd(), argv.out), JSON.stringify(result, null, 4));
  } else if (argv.format === FORMAT.BROWSER) {
    startServer(argv.port, result);
  } else if (argv.format === FORMAT.JSON) {
    await writeFileExtra(path.resolve(process.cwd(), argv.out), JSON.stringify(result, null, 4));
  } else if (argv.format === FORMAT.HTML) {
    await writeHTML(path.resolve(process.cwd(), argv.out), result);
  }

  console.log('finished analyzing.');
};

export {
  analyze,
  getImportDeclarationTree,
  fileCounter,
};
