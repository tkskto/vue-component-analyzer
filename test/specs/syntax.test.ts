import {join} from 'path';
import {getImportDeclarationTree} from '../../src/server/Analyzer';
const fixturesDir = join(__dirname, '../fixture/');

describe('syntax test', () => {
  it('optionalChaining.vue', async () => {
    const filename = 'syntaxTest/optionalChaining';
    const declaration = await getImportDeclarationTree(join(fixturesDir, `${filename}.vue`), [], true);
    const json = JSON.stringify(declaration).slice(0);

    expect(json).toMatchSnapshot();
  });

  it('logicalAssignment.vue', async () => {
    const filename = 'syntaxTest/logicalAssignment';
    const declaration = await getImportDeclarationTree(join(fixturesDir, `${filename}.vue`), [], true);
    const json = JSON.stringify(declaration).slice(0);

    expect(json).toMatchSnapshot();
  });

  it('numericSeparators.vue', async () => {
    const filename = 'syntaxTest/numericSeparators';
    const declaration = await getImportDeclarationTree(join(fixturesDir, `${filename}.vue`), [], true);
    const json = JSON.stringify(declaration).slice(0);

    expect(json).toMatchSnapshot();
  });

  it('classDefined.vue', async () => {
    const filename = 'syntaxTest/classDefined';
    const declaration = await getImportDeclarationTree(join(fixturesDir, `${filename}.vue`), [], true);
    const json = JSON.stringify(declaration).slice(0);

    expect(json).toMatchSnapshot();
  });

  it('langIsTypeScript.vue', async () => {
    const filename = 'syntaxTest/langIsTypeScript';
    const declaration = await getImportDeclarationTree(join(fixturesDir, `${filename}.vue`), [], true);

    expect(declaration).toStrictEqual({
      name: '/test/fixture/syntaxTest/langIsTypeScript.vue',
      props: '',
      size: 192,
      lastModifiedTime: 0,
      children: [],
    });
  });

  it('langIsTypeScriptWithSetup.vue', async () => {
    const filename = 'syntaxTest/langIsTypeScriptWithSetup';
    const declaration = await getImportDeclarationTree(join(fixturesDir, `${filename}.vue`), [], true);

    expect(declaration).toStrictEqual({
      name: '/test/fixture/syntaxTest/langIsTypeScriptWithSetup.vue',
      props: '',
      size: 146,
      lastModifiedTime: 0,
      children: [],
    });
  });
});
