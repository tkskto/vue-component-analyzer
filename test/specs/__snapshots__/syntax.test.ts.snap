// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syntax test classDefined.vue 1`] = `"{"name":"/test/fixture/syntaxTest/classDefined.vue","props":"","size":261,"lastModifiedTime":0,"children":[]}"`;

exports[`syntax test logicalAssignment.vue 1`] = `"{"name":"/test/fixture/syntaxTest/logicalAssignment.vue","props":"","size":219,"lastModifiedTime":0,"children":[]}"`;

exports[`syntax test numericSeparators.vue 1`] = `"{"name":"/test/fixture/syntaxTest/numericSeparators.vue","props":"","size":135,"lastModifiedTime":0,"children":[]}"`;

exports[`syntax test optionalChaining.vue 1`] = `"{"name":"/test/fixture/syntaxTest/optionalChaining.vue","props":"props: {\\n    data: {\\n        type: Object,\\n        required: true,\\n    },\\n}","size":286,"lastModifiedTime":0,"children":[]}"`;
